### PR TITLE
Deopt Underscore methods that change on arguments

### PR DESCRIPTION
**Pull request by jridgewell**
All of our currently deoptimized Underscore methods (`pick`, `omit`, `invoke`, `without`, and `difference`) change behavior based on the number of arguments pass to the call. This both adds `reduce` (and friends).

Supersedes (and incorporates) #3810.

### CODE REVIEW

File: backbone.js [hunk 1]
```diff
@@ -68,6 +68,9 @@
   // form param named `model`.
   Backbone.emulateJSON = false;
 
+  // A cached regex to detect if the Underscore function contains the `arguments`
+  // keyword.
+  var containsArguments = /arguments/;
```
Comment (akre54) [inline backbone.js]: this has to be `/\barguments\b/`, no?

File: backbone.js [hunk 2]
```diff
@@ -76,6 +79,10 @@
   //
   // `Function#apply` can be slow so we use the method's arg count, if we know it.
   var addMethod = function(length, method, attribute) {
+    // In case we don't catch it, deopt Underscore functions that contain
+    // the arguments keyword.
+    if (length > 0 && containsArguments.test(_[method])) length = 0;
```
Comment (akre54) [inline backbone.js]: I didn't know `Re::test` automatically `toString`'d functions. Neat.

**Comment by akre54** [comment]
This is a little gross but so is most of the perf rain dancing we're doing around those parts.

File: backbone.js (hunk 2 above)
Comment (jridgewell) [inline backbone.js]: Yup. It'll soon be configurable by one of the well-known symbols.

File: backbone.js (hunk 1 above)
Comment (jridgewell) [inline backbone.js]: Updated.

File: backbone.js (hunk 2 above)
Comment (akre54) [inline backbone.js]: eh?

Comment (jridgewell) [inline backbone.js]: Scratch that, I thought `#test` was included in http://www.2ality.com/2015/09/well-known-symbols-es6.html#forwarding_calls_from_string_methods_to_their_parameters.

**Comment by akre54** [comment]
>  explicitly deopts any method that depends on `arguments` from falling through the cracks

Do we really need to do this though?

**Comment by just-boris** [comment]
That is need because raw underscore methods behave so.
Why optimisation of `reduce`-family methods matters? This methods are applied for collection once per time and are not called so often

**Comment by jridgewell** [comment]
> Do we really need to do this though?

Necessarily, no. But'll it catch our screw-ups next time. And it might increase interop with underscore-like libraries that use `arguments` for other reasons.

**Comment by akre54** [comment]
> underscore-like libraries that use arguments for other reasons

The only library this could refer to is Lo-dash. And afaik he doesn't do anything with `arguments`. Unless I'm mistaken?

**Comment by jridgewell** [comment]
> The only library this could refer to is Lo-dash.

Not specifically, this could be any handwritten library.

> And afaik he doesn't do anything with arguments. Unless I'm mistaken?

I think lodash is pretty in-line with underscore on the functions Backbone uses.

**Comment by jridgewell** [comment]
Anyone else object?

File: backbone.js [hunk 3]
```diff
@@ -1166,8 +1173,8 @@
   // Underscore methods that we want to implement on the Collection.
   // 90% of the core usefulness of Backbone Collections is actually implemented
   // right here:
-  var collectionMethods = { forEach: 3, each: 3, map: 3, collect: 3, reduce: 4,
-      foldl: 4, inject: 4, reduceRight: 4, foldr: 4, find: 3, detect: 3, filter: 3,
+  var collectionMethods = { forEach: 3, each: 3, map: 3, collect: 3, reduce: 0,
```
Comment (akre54) [inline backbone.js]: Can you separate out these changes from the toString check? These should definitely get merged in.

Comment (jridgewell) [inline backbone.js]: Done.

**Pull request merged**